flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java

```diff
@@ -563,7 +590,14 @@ private void emitWindowContents(W window, ACC contents) throws Exception {
 	 * of the given window.
 	 */
 	protected boolean isLate(W window) {
-		return (windowAssigner.isEventTime() && (cleanupTime(window) <= internalTimerService.currentWatermark()));
+		if (windowAssigner.isEventTime()) {
+			if (windowAssigner instanceof GlobalEventTimeRowWindowAssigner) {
+				return windowAssignerContext.getCurrentElementTime() < windowAssignerContext.getCurrentMaxTime();
```

Comment:
I think this logic does not work very well if the data arrives is out of order.  How do you think?